ocaml/parsing/jane_syntax_parsing.ml

```diff
@@ -536,29 +673,91 @@ module Make_with_attribute
        include AST_syntactic_category
 
        val attributes : ast -> attributes
-       val with_attributes : ast -> attributes -> ast
-     end) : AST_internal with type ast = AST_syntactic_category.ast
+       val set_attributes : ast -> attributes -> ast
+     end) : AST_with_attributes_internal
+              with type ast = AST_syntactic_category.ast
 = struct
     include AST_syntactic_category
 
+    let add_attributes attrs ast = set_attributes ast (attributes ast @ attrs)
+
     let embedding_syntax = Embedding_syntax.Attribute
 
+    let make_attr loc name =
+      let loc = Location.ghostify loc in
+      { attr_name = { txt = Embedded_name.to_string name; loc }
+      ; attr_loc = loc
+      ; attr_payload = PStr []
+      }
+
     let make_jane_syntax name ast =
-      let attr =
-        { attr_name =
-            { txt = Embedded_name.to_string name
-            ; loc = !Ast_helper.default_loc
-            }
-        ; attr_loc = !Ast_helper.default_loc
-        ; attr_payload = PStr []
-        }
+      let attr = make_attr !Ast_helper.default_loc name in
+      let attrs =
+        match Embedded_name.components name with
+        | [feature_component] ->
+            (* Outermost; save the location *)
+            let save_loc = location ast in
+            [ make_attr
+                save_loc
+                { name
+                  with components =
+                         [ feature_component
+                         ; "_location"
+                         ; if save_loc.loc_ghost then "_ghost" else "_nonghost"]
+                }
+            ; attr ]
```

Comment:
Yes, we are exactly in want of that bit, unfortunately.  In re. your thoughts:

* I definitely think adding a motivated example is a good idea.

* `pexp_loc_stack` does sound promising, but I also have absolutely no idea what it is.  Not a dumb question at all!

* I thought about that but it seemed like a headache to code and the user would never see it.  If you think it would be an improvement, I can reevaluate that.

Reply:
We will revisit this after asking more broadly about `pexp_loc_stack`.

Reply:
`loc_stack` was exactly correct!  Please review the new `loc_stack` based code instead of this.  Thanks for the suggestion, this is at least 50x better.

Reply:
I like how this is looking. As we just discussed, I wonder if we can make `{Expression,Pattern,...}.make_entire_jane_syntax` always do this saving, and then make `{Expression,Pattern,...}.make_of_ast` always do the restoring before handing the ast off to `of_ast_internal`.

Reply:
Antal has now done this in afefad0b9019c02fec1aa4f224df82c5cb77d7cb.